File: pyInit/database_create.pl
#! /usr/bin/perl

# cpan DBD::mysql

use warnings;
use strict;
use FindBin;
use lib $FindBin::Bin;

use DBI;
use params;

#TODO: Delete the old database file
#TODO: Database file location better handled
our $dbhandle;

my $count_args = @ARGV ;

if ($count_args == 1) {
	my $db = $ARGV[0];
	if ($db =~ /db$/){
		print "valid";
		$params::database = $db;
	}
} 


dbConnect();
dropTables();
create_photo_table();
create_people_table();
create_linker_table();
create_root_dir_table();
create_metadata_table();
create_comments_table();

update_metadata();

# Vacuum the database
$dbhandle->do('VACUUM');
$dbhandle->disconnect;

# End creation of tables.

sub dbConnect{
	$dbhandle = DBI->connect("dbi:SQLite:$params::database", "user" , "pass");
}

sub dropTables{
	# Clean out the old tables so we can create them afresh.
	my $dropPhotos = qq/DROP TABLE IF EXISTS $params::photoTableName/;
	my $query = $dbhandle->prepare($dropPhotos);
	$query->execute() or die $DBI::errstr;

	my $dropPeople = qq/DROP TABLE IF EXISTS $params::peopleTableName/;
	$query = $dbhandle->prepare($dropPeople);
	$query->execute() or die $DBI::errstr;

	my $dropLinker = qq/DROP TABLE IF EXISTS $params::linkerTableName/;
	$query = $dbhandle->prepare($dropLinker);
	$query->execute() or die $DBI::errstr;

	my $dropRootDirs = qq/DROP TABLE IF EXISTS $params::rootTableName/;
	$query = $dbhandle->prepare($dropRootDirs);
	$query->execute() or die $DBI::errstr;

	my $dropMetadata = qq/DROP TABLE IF EXISTS $params::metadataTableName/;
	$query = $dbhandle->prepare($dropMetadata);
	$query->execute() or die $DBI::errstr;

	my $dropCommentLinker = qq/DROP TABLE IF EXISTS $params::commentLinkerUserTableName/;
	$query = $dbhandle->prepare($dropCommentLinker);
	$query->execute() or die $DBI::errstr;

	$dropCommentLinker = qq/DROP TABLE IF EXISTS $params::commentLinkerGoogleTableName/;
	$query = $dbhandle->prepare($dropCommentLinker);
	$query->execute() or die $DBI::errstr;

	$dropCommentLinker = qq/DROP TABLE IF EXISTS $params::commentLinkerClarifaiTableName/;
	$query = $dbhandle->prepare($dropCommentLinker);
	$query->execute() or die $DBI::errstr;

	
}

###############################################
########### subs for table creation ###########
###############################################

# Create the photo primary key and filename table
sub create_photo_table{
	# Create the database table, where photo_key is the primary key and photo_file is the file name 
	my $sql_quer = qq/CREATE TABLE $params::photoTableName (
	    $params::photoKeyColumn  INTEGER PRIMARY KEY AUTOINCREMENT
	                      UNIQUE,
	    $params::photoFileColumn STRING,
	    $params::photoDateColumn STRING,

	    $params::photoYearColumn STRING,
	    $params::photoMonthColumn STRING,
	    $params::photoDayColumn STRING,
	    $params::photoHourColumn STRING,
	    $params::photoMinuteColumn STRING,
	    $params::photoGMTColumn STRING,

	    $params::modifyDateColumn STRING,
	    $params::insertDateColumn STRING,
	    $params::rootDirNumColumn STRING,

	    $params::houseNumColumn STRING,
	    $params::streetColumn STRING,
	    $params::cityColumn STRING,
	    $params::stateColumn STRING,
	    $params::postcodeCoulumn STRING,
	    $params::countryColumn STRING,
	    $params::latColumn STRING,
	    $params::longColumn STRING
	); /;

	# Prepare and execute the statement
	my $sub_state_handle = $dbhandle->prepare($sql_quer);
	$sub_state_handle->execute() or die $DBI::errstr;
}

sub create_comments_table{
	my $sql_quer = qq/CREATE TABLE $params::commentLinkerUserTableName (
	    $params::commentLinkerPhotoColumn INTEGER REFERENCES $params::photoTableName ($params::photoKeyColumn),
	    $params::commentLinkerTagColumn STRING,
	    $params::commentLinkerTagProbabilityColumn DOUBLE
	); /;
	
	my $sub_state_handle = $dbhandle->prepare($sql_quer);
	$sub_state_handle->execute() or die $DBI::errstr;


	$sql_quer = qq/CREATE TABLE $params::commentLinkerGoogleTableName (
	    $params::commentLinkerPhotoColumn INTEGER REFERENCES $params::photoTableName ($params::photoKeyColumn),
	    $params::commentLinkerTagColumn STRING,
	    $params::commentLinkerTagProbabilityColumn DOUBLE
	); /;
	
	$sub_state_handle = $dbhandle->prepare($sql_quer);
	$sub_state_handle->execute() or die $DBI::errstr;


	$sql_quer = qq/CREATE TABLE $params::commentLinkerClarifaiTableName (
	    $params::commentLinkerPhotoColumn INTEGER REFERENCES $params::photoTableName ($params::photoKeyColumn),
	    $params::commentLinkerTagColumn STRING,
	    $params::commentLinkerTagProbabilityColumn DOUBLE
	); /;
	
	$sub_state_handle = $dbhandle->prepare($sql_quer);
	$sub_state_handle->execute() or die $DBI::errstr;
}

sub create_root_dir_table{
	my $sql_quer = qq/CREATE TABLE $params::rootTableName (
		$params::rootKeyColumn INTEGER PRIMARY KEY AUTOINCREMENT UNIQUE,
		$params::windowsRootPath STRING,
		$params::linuxRootPath STRING
	); /;

	my $sub_state_handle = $dbhandle->prepare($sql_quer);
	$sub_state_handle->execute() or die $DBI::errstr;
}

# Create the people table
sub create_people_table{
	
	# Create the database table, where photo_key is the primary key and photo_file is the file name 
	my $sql_quer = qq/CREATE TABLE $params::peopleTableName (
	    $params::peopleKeyColumn  INTEGER PRIMARY KEY AUTOINCREMENT
	                            UNIQUE,
	    $params::personNameColumn STRING
	);/;

	#,$params::personPicasaID BIGINT

	my $sub_state_handle = $dbhandle->prepare($sql_quer);
	$sub_state_handle->execute() or die $DBI::errstr;

	my $ben_query = qq/INSERT INTO $params::peopleTableName ($params::personNameColumn) VALUES ("Benjamin Lewis");/;
	my $jessica_query = qq/INSERT INTO $params::peopleTableName ($params::personNameColumn) VALUES ("Jessica Lewis");/;

	$sub_state_handle = $dbhandle->prepare($ben_query);
	$sub_state_handle->execute() or die $DBI::errstr;

	$sub_state_handle = $dbhandle->prepare($jessica_query);
	$sub_state_handle->execute() or die $DBI::errstr;
}

# Create the linkage table
sub create_linker_table{
	my $sql_quer = qq/CREATE TABLE $params::linkerTableName (
	    $params::linkerPhotoColumn  INTEGER REFERENCES $params::photoTableName ($params::photoKeyColumn),
	    $params::linkerPeopleColumn INTEGER REFERENCES $params::peopleTableName ($params::peopleKeyColumn) 
	);/;

	my $sub_state_handle = $dbhandle->prepare($sql_quer);
	$sub_state_handle->execute() or die $DBI::errstr;

}

sub create_metadata_table{
	my $sql_quer = qq/CREATE TABLE $params::metadataTableName (
		$params::metadataNameColumn  STRING,  
		$params::metadataValueColumn STRING
	);/;

	my $metadata_handle = $dbhandle->prepare($sql_quer);
	$metadata_handle->execute() or die $DBI::errstr;

}

sub update_metadata{

	my ($sec,$min,$hour,$mday,$mon,$year) = localtime(time);
	$year += 1900;
	$mon += 1;

	my $dateTime = sprintf("%04d-%02d-%02d %02d:%02d:%02d", $year, $mon, $mday, $hour, $min, $sec);

	my $updatedValQuery = qq/INSERT INTO $params::metadataTableName VALUES ("$params::metadataLastEditedField", "$dateTime")/;
	
	my $metadata_handle = $dbhandle->prepare($updatedValQuery);
	$metadata_handle->execute() or die $DBI::errstr;

}

File: pyInit/database_create.py
#! /usr/bin/env python

import sqlite3
from sqlite3 import Error
import os
import xmltodict
import time

project_path = os.path.abspath(os.path.join(__file__,"../.."))
script_path  = os.path.abspath(os.path.join(__file__,".."))

with open(os.path.join(project_path, 'config/params.xml') ) as stream:
    try:
        params = xmltodict.parse(stream.read())
    except Exception as exc:
        print(exc)
        exit(1)


# People table parameters
peopleTableName = params['params']['photoDatabase']['tables']['peopleTable']['Name']
peopleKeyColumn = params['params']['photoDatabase']['tables']['peopleTable']['Columns']['peopleKey']
personNameColumn = params['params']['photoDatabase']['tables']['peopleTable']['Columns']['personName']

# Linker table parameters
linkerTable = params['params']['photoDatabase']['tables']['photoLinkerTable']
linkerTableName = linkerTable['Name']
linkerPhotoColumn = linkerTable['Columns']['linkerPhoto']
linkerPeopleColumn = linkerTable['Columns']['linkerPeople']

# Root table parameters
rootTableName = params['params']['photoDatabase']['tables']['rootTable']['Name']
rootKeyColumn = params['params']['photoDatabase']['tables']['rootTable']['Columns']['rootKey']
windowsRootPath = params['params']['photoDatabase']['tables']['rootTable']['Columns']['windowsRootPath']
linuxRootPath = params['params']['photoDatabase']['tables']['rootTable']['Columns']['linuxRootPath']

# Metadata table parameters
metadataTableName = params['params']['photoDatabase']['tables']['metadataTable']['Name']
metadataNameColumn = params['params']['photoDatabase']['tables']['metadataTable']['Columns']['metadataName']
metadataValueColumn = params['params']['photoDatabase']['tables']['metadataTable']['Columns']['metadataValue']
metadataLastEditedField = params['params']['photoDatabase']['tables']['metadataTable']['Fields']['metadataLastEditedField']

# Master table of comments/tags -- for deduplication
masterCommentTableName = params['params']['photoDatabase']['tables']['masterComments']['Name']
masterCommentTagIDNum = params['params']['photoDatabase']['tables']['masterComments']['Columns']['tagID']
masterCommentTagString = params['params']['photoDatabase']['tables']['masterComments']['Columns']['tagString']

# Comment linker table parameters -- user
commentLinkerUserTableName = params['params']['photoDatabase']['tables']['commentLinkerUserTable']['Name']
commentLinkerPhotoColumnUser = params['params']['photoDatabase']['tables']['commentLinkerUserTable']['Columns']['commentLinkerPhoto']
commentLinkerTagColumnUser = params['params']['photoDatabase']['tables']['commentLinkerUserTable']['Columns']['commentLinkerTagIDlink']
commentLinkerTagProbabilityColumnUser = params['params']['photoDatabase']['tables']['commentLinkerUserTable']['Columns']['commentLinkerTagProbability']

# Comment linker table parameters -- machine generated 
commentLinkerMachineTableName = params['params']['photoDatabase']['tables']['machineCommentLinker']['Name']
commentLinkerPhotoColumnMachine = params['params']['photoDatabase']['tables']['machineCommentLinker']['Columns']['commentLinkerPhoto']
commentLinkerTagColumnMachine = params['params']['photoDatabase']['tables']['machineCommentLinker']['Columns']['commentLinkerTagIDlink']
commentLinkerTagProbabilityColumnMachine = params['params']['photoDatabase']['tables']['machineCommentLinker']['Columns']['commentLinkerTagProbability']

# commentLinkerGoogleTableName = params['params']['photoDatabase']['tables']['commentLinkerGoogleTable']['Name']
# commentLinkerPhotoColumnGoogle = params['params']['photoDatabase']['tables']['commentLinkerGoogleTable']['Columns']['commentLinkerPhoto']
# commentLinkerTagColumnGoogle = params['params']['photoDatabase']['tables']['commentLinkerGoogleTable']['Columns']['commentLinkerTag']
# commentLinkerTagProbabilityColumnGoogle = params['params']['photoDatabase']['tables']['commentLinkerGoogleTable']['Columns']['commentLinkerTagProbability']

# commentLinkerClarifaiTableName = params['params']['photoDatabase']['tables']['commentLinkerClarifaiTable']['Name']
# commentLinkerPhotoColumnClarifai = params['params']['photoDatabase']['tables']['commentLinkerClarifaiTable']['Columns']['commentLinkerPhoto']
# commentLinkerTagColumnClarifai = params['params']['photoDatabase']['tables']['commentLinkerClarifaiTable']['Columns']['commentLinkerTag']
# commentLinkerTagProbabilityColumnClarifai = params['params']['photoDatabase']['tables']['commentLinkerClarifaiTable']['Columns']['commentLinkerTagProbability']



# Photo table Parameters
photoTableParams = params['params']['photoDatabase']['tables']['photoTable']
phtp = photoTableParams # shorter, but I want it to be clear what it means. 

photoTableName = phtp['Name']
photoTableColumns = phtp['Columns']
phtc = photoTableColumns

photoKeyColumn = phtc['photoKey']
photoFileColumn = phtc['photoFile']
photoDateColumn = phtc['photoDate']

photoYearColumn = phtc['photoYear']
photoMonthColumn = phtc['photoMonth']
photoDayColumn = phtc['photoDay']
photoHourColumn = phtc['photoHour']
photoMinuteColumn = phtc['photoMinute']
photoGMTColumn = phtc['photoGMT']

modifyDateColumn = phtc['modifyDate']
insertDateColumn = phtc['insertDate']
rootDirNumColumn = phtc['rootDirNum']

houseNumColumn = phtc['houseNum']
streetColumn = phtc['street']
cityColumn = phtc['city']
stateColumn = phtc['state']
postcodeCoulumn = phtc['postcode']
countryColumn = phtc['country']
latColumn = phtc['lat']
longColumn = phtc['long']
# End photo table parameters



def connectDB(params):

    dbName = params['params']['photoDatabase']['fileName']

    conn = sqlite3.connect(os.path.join(project_path, 'databases', dbName))
    conn.text_factory = lambda x: unicode(x, "utf-8", "ignore")
    return conn

def dropTables(conn):
    """ create a database connection to a SQLite database """
    cc = conn.cursor()
    dropPhotos = 'DROP TABLE IF EXISTS ' + photoTableName
    cc.execute(dropPhotos)

    dropPeople = 'DROP TABLE IF EXISTS ' + peopleTableName
    cc.execute(dropPeople)

    dropLinker = 'DROP TABLE IF EXISTS ' + linkerTableName
    cc.execute(dropLinker)

    dropRootDirs = 'DROP TABLE IF EXISTS ' + rootTableName
    cc.execute(dropRootDirs)

    dropMetadata = 'DROP TABLE IF EXISTS ' + metadataTableName
    cc.execute(dropMetadata)

    dropCommentLinker = 'DROP TABLE IF EXISTS ' + commentLinkerUserTableName
    cc.execute(dropCommentLinker)

    dropCommentLinker = 'DROP TABLE IF EXISTS ' + commentLinkerMachineTableName
    cc.execute(dropCommentLinker)

    # Must be dropped after the comment linker child tables. 
    dropCommentLinker = 'DROP TABLE IF EXISTS ' + masterCommentTableName
    cc.execute(dropCommentLinker)

    conn.commit()

# # Create the photo primary key and filename table
def create_photo_table(conn):
    # Create the database table, where photo_key is the primary key and photo_file is the file name


    sql_quer = 'CREATE TABLE ' + \
        photoTableName + '(' + \
        photoKeyColumn + '  INTEGER PRIMARY KEY AUTOINCREMENT UNIQUE, ' + \
        photoFileColumn + ' STRING, ' + \
        photoDateColumn + ' STRING, ' + \
                                        \
        photoYearColumn + ' STRING, ' + \
        photoMonthColumn + ' STRING, ' + \
        photoDayColumn + ' STRING, ' + \
        photoHourColumn + ' STRING, ' + \
        photoMinuteColumn + ' STRING, ' + \
        photoGMTColumn + ' STRING, ' + \
                                        \
        modifyDateColumn + ' STRING, ' + \
        insertDateColumn + ' STRING, ' + \
        rootDirNumColumn + ' STRING, ' + \
                                        \
        houseNumColumn + ' STRING, ' + \
        streetColumn + ' STRING, ' + \
        cityColumn + ' STRING, ' + \
        stateColumn + ' STRING, ' + \
        postcodeCoulumn + ' STRING, ' + \
        countryColumn + ' STRING, ' + \
        latColumn + ' STRING, ' + \
        longColumn + ' STRING ); '

    c = conn.cursor()
    c.execute(sql_quer)
    conn.commit()

# Create the people table
def create_people_table(conn):
    

    # Create the database table, where photo_key is the primary key and photo_file is the file name 
    sql_quer = 'CREATE TABLE ' + peopleTableName + ' ( ' + \
        peopleKeyColumn + ' INTEGER PRIMARY KEY AUTOINCREMENT UNIQUE, ' + \
        personNameColumn + ' STRING  );'

    c = conn.cursor()
    c.execute(sql_quer)
    conn.commit()

    ben_query = 'INSERT INTO ' + peopleTableName + '(' + personNameColumn + ') VALUES ("Benjamin Lewis");'
    jessica_query = 'INSERT INTO ' + peopleTableName + '(' + personNameColumn + ') VALUES ("Jessica Lewis");'

    c.execute(ben_query)
    c.execute(jessica_query)
    conn.commit()

# Create the linkage table
def create_linker_table(conn):


    sql_quer = 'CREATE TABLE ' + linkerTableName + ' ( ' + \
        linkerPhotoColumn  + ' INTEGER REFERENCES ' + photoTableName + ' ( ' + photoKeyColumn  + ' ), ' + \
        linkerPeopleColumn + ' INTEGER REFERENCES ' + peopleTableName + ' ( ' + peopleKeyColumn  + ' ) );'

    c = conn.cursor()
    c.execute(sql_quer)
    conn.commit()

def create_root_dir_table(conn):
    sql_quer = 'CREATE TABLE ' + rootTableName + ' ( ' \
        + rootKeyColumn + ' INTEGER PRIMARY KEY AUTOINCREMENT UNIQUE, ' \
        + windowsRootPath + ' STRING, ' \
        + linuxRootPath + ' STRING ); ' 
     
    c = conn.cursor()
    c.execute(sql_quer)
    conn.commit()


def create_metadata_table(conn):
    sql_quer = 'CREATE TABLE ' + metadataTableName + ' ('\
        + metadataNameColumn + '  STRING,  '\
        + metadataValueColumn + ' STRING);'

    c = conn.cursor()
    c.execute(sql_quer)
    conn.commit()

def create_comments_table(conn):


# # Master table of comments/tags -- for deduplication
# masterCommentTableName = params['params']['photoDatabase']['tables']['commentLinkerUserTable']['masterComments']['Name']
# masterCommentTagIDNum = params['params']['photoDatabase']['tables']['commentLinkerUserTable']['masterComments']['Columns']['tagID']
# masterCommentTagString = params['params']['photoDatabase']['tables']['commentLinkerUserTable']['masterComments']['Columns']['tagString']

# # Comment linker table parameters -- user
# commentLinkerUserTableName = params['params']['photoDatabase']['tables']['commentLinkerUserTable']['Name']
# commentLinkerPhotoColumnUser = params['params']['photoDatabase']['tables']['commentLinkerUserTable']['Columns']['commentLinkerPhoto']
# commentLinkerTagColumnUser = params['params']['photoDatabase']['tables']['commentLinkerUserTable']['Columns']['commentLinkerTagIDlink']
# commentLinkerTagProbabilityColumnUser = params['params']['photoDatabase']['tables']['commentLinkerUserTable']['Columns']['commentLinkerTagProbability']

# # Comment linker table parameters -- machine generated 
# commentLinkerMachineTableName = params['params']['photoDatabase']['tables']['machineCommentLinker']['Name']
# commentLinkerPhotoColumnMachine = params['params']['photoDatabase']['tables']['machineCommentLinker']['Columns']['commentLinkerPhoto']
# commentLinkerTagColumnMachine = params['params']['photoDatabase']['tables']['machineCommentLinker']['Columns']['commentLinkerTagIDlink']
# commentLinkerTagProbabilityColumnMachine = params['params']['photoDatabase']['tables']['machineCommentLinker']['Columns']['commentLinkerTagProbability']

    sql_quer = 'CREATE TABLE ' + masterCommentTableName + ' (' + \
        masterCommentTagIDNum + '  INTEGER PRIMARY KEY AUTOINCREMENT UNIQUE, ' + \
        masterCommentTagString + ' STRING);'

    c = conn.cursor()
    c.execute(sql_quer)
    conn.commit()


    sql_quer = 'CREATE TABLE ' + commentLinkerUserTableName + ' (' + \
        commentLinkerPhotoColumnUser + ' INTEGER REFERENCES ' + photoTableName + ' ( ' + photoKeyColumn + '), ' + \
        commentLinkerTagColumnUser + ' INTEGER REFERENCES  ' + masterCommentTableName + ' ( ' + masterCommentTagIDNum + '), ' + \
        commentLinkerTagProbabilityColumnUser + ' DOUBLE);'
    
    c.execute(sql_quer)
    conn.commit()


    sql_quer = 'CREATE TABLE ' + commentLinkerMachineTableName + ' (' + \
        commentLinkerPhotoColumnMachine + ' INTEGER REFERENCES ' + photoTableName + ' ( ' + photoKeyColumn + '), ' + \
        commentLinkerTagColumnMachine + ' INTEGER REFERENCES  ' + masterCommentTableName + ' ( ' + masterCommentTagIDNum + '), ' + \
        commentLinkerTagProbabilityColumnMachine + ' DOUBLE);'
    
    c.execute(sql_quer)
    conn.commit()


def update_metadata(conn):

    # (year, month, day, hour, minute, second, wday, yearday, isdst) = time.localtime()

    # my $dateTime = sprintf("%04d-%02d-%02d %02d:%02d:%02d", $year, $mon, $mday, $hour, $min, $sec);
    dateTime = time.strftime('%Y-%m-%d %H:%M:%S', time.localtime())

    updatedValQuery = 'INSERT INTO ' + metadataTableName + ' VALUES ("' + metadataLastEditedField + '", "' + dateTime + '")'

    c = conn.cursor()
    c.execute(updatedValQuery)
    conn.commit()


conn = connectDB(params)
dropTables(conn)
create_photo_table(conn)
create_people_table(conn)
create_linker_table(conn)
create_root_dir_table(conn)
create_metadata_table(conn)
create_comments_table(conn)

update_metadata(conn)
